main.go

```diff
@@ -3,45 +3,147 @@ package main
+	StagingBucket string
+
+	// final bucket for firehose files
+	OutputBucket string
 }
 
-// ReceivedMessage is the message as received by a transport backend.
 type ReceivedMessage struct {
-	Payload          []byte
-	Attributes       map[string]string
-	ProviderMetadata interface{}
+	message *hedwig.Message
+	errCh   chan error
+}
+
+// StorageBackendCreator is used for read/write to storage
+type StorageBackendCreator interface {
+	CreateWriter(ctx context.Context, uploadBucket string, uploadLocation string) (io.WriteCloser, error)
+	CreateReader(ctx context.Context, uploadBucket string, uploadLocation string) (io.ReadCloser, error)
 }
 
-// StorageBackend is used for read/write to storage
-type StorageBackend interface {
-	UploadFile(ctx context.Context, data []byte, uploadBucket string, uploadLocation string) error
+type Clock struct {
+	instant time.Time
+}
 
-	ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error)
+func (this *Clock) Now() time.Time {
+	if this == nil {
+		return time.Now()
+	}
+	return this.instant
 }
 
 type Firehose struct {
-	processSettings ProcessSettings
-	storageBackend  StorageBackend
-	hedwigConsumer  *hedwig.QueueConsumer
+	processSettings       ProcessSettings
+	storageBackendCreator StorageBackendCreator
+	hedwigConsumer        *hedwig.QueueConsumer
+	hedwigFirehose        *hedwig.Firehose
+	flushLock             sync.Mutex
+	flushCh               chan error
+	messageCh             chan ReceivedMessage
+	listenRequest         hedwig.ListenRequest
+	clock                 *Clock
+}
+
+func (fp *Firehose) flushCron(ctx context.Context) {
+	errChannelMapping := make(map[hedwig.MessageTypeMajorVersion][]chan error)
+	writerMapping := make(map[hedwig.MessageTypeMajorVersion]io.WriteCloser)
+	currentTime := fp.clock.Now()
+	timerCh := time.After(time.Duration(fp.processSettings.FlushAfter) * time.Second)
+	// go through all msgs and write to msgtype folder
+	for {
+		select {
+		case <-timerCh:
+			// close all writers and associated errChannels
+			for key, writer := range writerMapping {
+				err := writer.Close()
+				errChannels := errChannelMapping[key]
+				if err != nil {
+					for _, errCh := range errChannels {
+						errCh <- err
+					}
+				} else {
+					for _, errCh := range errChannels {
+						close(errCh)
+					}
+				}
+			}
+			// start a new flushcron go routine, as this one is done
+			go fp.flushCron(ctx)
+			return
+		case messageAndChan := <-fp.messageCh:
+			message := messageAndChan.message
+			errCh := messageAndChan.errCh
+			key := hedwig.MessageTypeMajorVersion{
+				MessageType:  message.Type,
+				MajorVersion: uint(message.DataSchemaVersion.Major()),
+			}
+			// if writer doesn't exist create in mapping
+			if _, ok := writerMapping[key]; !ok {
+				uploadLocation := fmt.Sprintf("%s/%s/%s/%s", key.MessageType, fmt.Sprint(key.MajorVersion), currentTime.Format("2006/1/2"), fmt.Sprint(currentTime.Unix()))
```

Comment:
Is the hope that there's no conflicts here because we're using epoch timestamp? Should this also include a "node id" to be more explicit?

Reply:
ah I see what you mean here I guess there is the chance that the timestamp could conflict - how would I generate the node id?

Reply:
ok will add this once node startup / leader election code is added